### PR TITLE
DYNCFG: dynamically configured alerts

### PR DESCRIPTION
**Pull request by ktsaou**
Currently doing preparative work to enable dynamic configuration.

- [x] health config files are now loaded as RRD_ALERT_PROTOTYPE. This structure is suitable for both RRDCALC and RRDCALCTEMPLATE.
- [x] once the prototypes have been loaded, they are applied to individual hosts from memory. No need to load them from disk again.
- [x] prototypes now contain an **array of rules**, each containing 2 sub-structures: **match** and **config**. Match contains all the attributes for matching alerts to hosts and instances. Config includes all the configuration of the alert.
- [x] DYNCFG requires unique IDs for everything configurable. There are 2 levels at which a configuration can be made:
   - **global prototypes**:<br/> these configurations are applied to all nodes, both templates and individual alerts.
   - **alert instances**:</br> these are always individual instances of the alerts.

   Using this setup, a user will be able to configure an alert to be applied to all instances on all nodes (like editing a file), to all instances of a specific node, to a specific instance of a specific node (so, fine tune each specific alert).

   When multiple alerts are read from the configuration file, they are added as rules to the same alert. So, when multiple overlapping alerts exist, the prototypes have an entry for each alert name, having a linked list of all the rules that have been configured for it.

   Dynamic configuration will take place at the alert as a whole, for all its rules.

- [x] removed old indexes of alert variables (rrddimvar, rrdsetvar, rrdfamily)
- [x] removed rrdcalctemplate 
- [x] the remaining RRDVAR has be tremendously simplified
- [x] introduced a new function to lookup alert variables.
- [x] lookup of alert variables now scores the instances to find the most suitable one for each instance.
- [x] new `/api/v1/variable?chart=X&variable=Y` allows examining operation of the alert variable lookup.
- [x] Alerts configuration hashing now happens with XXH3 128bits
- [x] Foreach dimension function has been removed
- [x] fixed 2 memory leaks, one if the loaded RRDCALC is not used (it was not freed), another about `summary` members that were never freed.
- [x] cleaned up shadow variables and compiler warnings so that netdata compiles without any warning with gcc/clang flags: ` -Wall -Wextra -Wformat=2 -Wshadow -Wno-format-nonliteral -Winit-self`

---

For users: the changes in this PR may introduce a slightly higher CPU load for alert variables resolution, but the memory reduction is vast.

### CODE REVIEW

**Comment by ktsaou** [comment]
I think it would be nice to merge this.
Dynamic configuration for alerts is still missing 3 things:

1. Parsing the JSON payload to alter the configuration of an alarm.
2. Work on the schema to of the alerts to provide a useful UI for editing them.
3. Integrating prior work on testing alerts before applying them.

**Comment by ktsaou** [comment]
@stelfrag alert transitions on the UI in many cases they don't show up. Something is wrong with the SQL code I think...

**Comment by ktsaou** [comment]
CI fails because of this:

```
cc1: error: '-Wformat-security' ignored without '-Wformat' [-Werror=format-security]
cc1: some warnings being treated as errors
```

@vkalintiris @Ferroin @tkatsoulas

**Comment by vkalintiris** [comment]
> CI fails because of this:
> 
> ```
> cc1: error: '-Wformat-security' ignored without '-Wformat' [-Werror=format-security]
> cc1: some warnings being treated as errors
> ```
> 
> @vkalintiris @Ferroin @tkatsoulas

@Ferroin from the logs:

```
[38/509] /usr/bin/gcc -DHAVE_CONFIG_H -D_GNU_SOURCE -I../mqtt_websockets/c_rhash/include -I../mqtt_websockets/c-rbuf/include -I../mqtt_websockets/src/include -I../web/server/h2o/libh2o/include -I../web/server/h2o/libh2o/deps/cloexec -I../web/server/h2o/libh2o/deps/brotli/enc -I../web/server/h2o/libh2o/deps/golombset -I../web/server/h2o/libh2o/deps/libgkc -I../web/server/h2o/libh2o/deps/libyrmcds -I../web/server/h2o/libh2o/deps/klib -I../web/server/h2o/libh2o/deps/neverbleed -I../web/server/h2o/libh2o/deps/picohttpparser -I../web/server/h2o/libh2o/deps/picotest -I../web/server/h2o/libh2o/deps/yaml/include -I../web/server/h2o/libh2o/deps/yoml -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fexceptions -DNDEBUG   -ffunction-sections -fdata-sections -Wno-all -Wno-extra -Wno-shadow -Wno-deprecated-declarations -DH2O_USE_LIBUV=0 -std=gnu11 -MD -MT CMakeFiles/h2o.dir/web/server/h2o/libh2o/deps/cloexec/cloexec.c.o -MF CMakeFiles/h2o.dir/web/server/h2o/libh2o/deps/cloexec/cloexec.c.o.d -o CMakeFiles/h2o.dir/web/server/h2o/libh2o/deps/cloexec/cloexec.c.o -c ../web/server/h2o/libh2o/deps/cloexec/cloexec.c
```

The warnings are treated as errors for `-Werror=format-security`. This not added by CMake but by the way packages are built for RPMs (and maybe DEBs).

**Comment by tkatsoulas** [comment]
CC @Ferroin 
https://github.com/netdata/netdata/pull/16779/commits/01824f2023c193e2e129534b8dd340ac4f302587
do I miss something?

File: database/rrdcalc.h [hunk 1]
```diff
@@ -25,29 +35,42 @@ typedef enum {
...
+
+    SIMPLE_PATTERN *os_pattern;
+    SIMPLE_PATTERN *host_pattern;
+    SIMPLE_PATTERN *charts_pattern;         // the simple pattern of charts
+    SIMPLE_PATTERN *plugin_pattern;         // the simple pattern of plugin
+    SIMPLE_PATTERN *module_pattern;         // the simple pattern of module
+    SIMPLE_PATTERN *host_labels_pattern;    // the simple pattern of labels
+    SIMPLE_PATTERN *chart_labels_pattern;   // the simple pattern of chart labels
+};
+void rrd_alert_match_cleanup(struct rrd_alert_match *am);
```
Comment (vkalintiris) [inline database/rrdcalc.h:68]: This should be moved under `health/`.

File: database/rrdcalc.h [hunk 2]
```diff
@@ -25,29 +35,42 @@ typedef enum {
...
 
-    uuid_t config_hash_id;          // a predictable hash_id based on specific alert configuration
+struct rrd_alert_config {
```
Comment (vkalintiris) [inline database/rrdcalc.h:70]: I think this should be moved under `health` as well.

**Review by vkalintiris — COMMENTED** [review]
```
typedef struct dyncfg {
    RRDHOST *host;
    uuid_t host_uuid;
```

Why do you need to associate a configuration with a host instead of the other way around? This will lead to the same issues we had with health structs inside rrd.

File: database/rrdcalc.h (hunk 1 above)
Comment (ktsaou) [reply database/rrdcalc.h]: moved rrdcalc and rrdvar to health

File: database/rrdcalc.h (hunk 2 above)
Comment (ktsaou) [reply database/rrdcalc.h]: same as above

**Comment by ktsaou** [comment]
@vkalintiris RRDHOST removed from DYNCFG. Now it looks up the host every time.

**Review by stelfrag — APPROVED** [review]

**Pull request merged**